source/facter/3.0/release_notes.md

```diff
@@ -235,6 +235,18 @@ We changed the value of the hardware fact (`os.hardware`) from "x64" to "x86_64"
 
 * [FACT-610](https://tickets.puppetlabs.com/browse/FACT-610)
 
+### BREAK: Removed `file_read.rb` and `registry.rb`
+
+Facter 3 removed the [`file_read.rb`](https://github.com/puppetlabs/facter/blob/2.x/lib/facter/util/file_read.rb) and [`registry.rb`](https://github.com/puppetlabs/facter/blob/2.x/lib/facter/util/registry.rb), which provide the `Facter::Util::FileRead` and `Facter::Util::Registry` modules to Facter. This can break custom facts that rely on these modules, and produce error messages like this:
```

Comment:
IIUC the issue isn't specific to these two files. E.g. look at https://github.com/puppetlabs/facter/tree/2.x/lib/facter/util, none of which exist in facter 3. Thus, the reimplementation of facter from ruby to C++ means custom facts require'ing any individual file in the previous implementation will fail. 

In addition, whereas facter 2 effectively exposed almost its entire implementation as if it were public API for authors of custom facts, facter 3 supports a well-considered API: https://github.com/puppetlabs/facter/blob/master/Extensibility.md#custom-facts-compatibility. Thus, custom facts using facter 2 "API"s outside the facter 3 supported set will fail.

With that as background, I'd suggest that:
* we document the APIs in that Extensibility.md link
* this section be reworked to instruct custom facts authors not to require individual implementation files and to prefer only the documented APIs

Since that first point may take a bit of time we could just link to the Extensibility.md for the time being.

And then that second point could then be enhanced to say something like: if you are porting custom facts from facter 2 that rely on old APIs, a workaround would be to copy the facter 2 implementation into the module's `lib/facter` directory as  described below.

Hope that makes sense. Happy to chat more about this in meatspace.

Also /cc @peterhuene for his thoughts.